pkg/server/image_pull.go

```diff
@@ -26,195 +26,223 @@ import (
-			fmt.Errorf("size failed for image:%q %v", image.Target.Digest, err)
-	}
-	compressedSize = uint64(size)
-	return resolvedImageName, manifest, compressedSize, nil
+	return &runtime.PullImageResponse{ImageRef: imageID}, err
 }
 
-func (c *criContainerdService) pullImage(ctx context.Context, ref string) (imagespec.Descriptor, int64, error) {
-	var (
-		err               error
-		size              int64
-		desc              imagespec.Descriptor
-		resolvedImageName string
-		fetcher           remotes.Fetcher
-	)
-
-	// Resolve the image name; place that in the image store; then dispatch
-	// a handler for a sequence of handlers which: 1) fetch the object using a
-	// FetchHandler; and 3) recurse through any sub-layers via a ChildrenHandler
+// pullImage pulls image and returns image id (config digest) and digest.
+// TODO(random-liu): [P0] Wait for all downloadings to be done before return.
+func (c *criContainerdService) pullImage(ctx context.Context, ref string) (
+	imagedigest.Digest, imagedigest.Digest, error) {
+	// Resolve the image reference to get descriptor and fetcher.
 	resolver := docker.NewResolver()
-
-	resolvedImageName, desc, fetcher, err = resolver.Resolve(ctx, ref)
+	_, desc, fetcher, err := resolver.Resolve(ctx, ref)
 	if err != nil {
-		return desc, size, fmt.Errorf("failed to resolve ref %q: err: %v", ref, err)
+		return "", "", fmt.Errorf("failed to resolve ref %q: %v", ref, err)
 	}
+	// Currently, the resolved image name is the same with ref in docker resolver,
+	// but they may be different in the future.
+	// TODO(random-liu): Always resolve image reference and use resolved image name in
+	// the system.
 
-	err = c.imageStoreService.Put(ctx, resolvedImageName, desc)
-	if err != nil {
-		return desc, size, fmt.Errorf("failed to put %q: desc: %v err: %v", resolvedImageName, desc, err)
+	// Put the image information into containerd image store.
+	if err = c.imageStoreService.Put(ctx, ref, desc); err != nil {
+		return "", "", fmt.Errorf("failed to put image %q desc %v into containerd image store: %v",
+			ref, desc, err)
 	}
+	// TODO(random-liu): What if following operations fail? Do we need to do cleanup?
 
+	// Fetch all image resources into content store.
+	// Dispatch a handler for a sequence of handlers which:
+	// 1) fetch the object using a FetchHandler;
+	// 2) recurse through any sub-layers via a ChildrenHandler.
 	err = containerdimages.Dispatch(
 		ctx,
 		containerdimages.Handlers(
 			remotes.FetchHandler(c.contentIngester, fetcher),
 			containerdimages.ChildrenHandler(c.contentProvider)),
 		desc)
 	if err != nil {
-		return desc, size, fmt.Errorf("failed to fetch %q: desc: %v err: %v", resolvedImageName, desc, err)
+		return "", "", fmt.Errorf("failed to fetch image %q desc %+v: %v", ref, desc, err)
 	}
 
-	image, err := c.imageStoreService.Get(ctx, resolvedImageName)
+	// Read the image manifest from content store.
+	image, err := c.imageStoreService.Get(ctx, ref)
```

Comment:
@yujuhong Yeah, that's why they introduced the containerd client. Their api is combined with libraries and services, it's hard to consume directly.